Projections

amosbird opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add projection support for MergeTree* tables. 

Detailed description / Documentation draft:

The documentation is almost listed in  https://github.com/ClickHouse/ClickHouse/issues/14730

Possible TODOs.

- [x] Complete query analysis for projections.
- [x] Normal projections with different order by clause
- [ ] ~~ProjectionMergeTree to store projection data only~~ can be done via CollapsingMergeTree with default sign = 1
- [ ] Secondary indices via projection

alexey-milovidov commented:
> Some exception occurred :( 

`GithubException: 500 null`
This is GitHub downtime, https://www.githubstatus.com/history

Need to restart the tests.

amosbird committed:
Projections

TODO (suggested by Nikolai)

1. Build query plan fro current query (inside storage::read) up to WithMergableState
2. Check, that plan is simple enough: Aggregating - Expression - Filter - ReadFromStorage (or simplier)
3. Check, that filter is the same as filter in projection, and also expression calculates the same aggregation keys as in projection
4. Return WithMergableState if projection applies

3 will be easier to do with ActionsDAG, cause it sees all functions, and dependencies are direct (but it is possible with ExpressionActions also)

Also need to figure out how prewhere works for projections, and
row_filter_policies.

wip

zhaojintaozhao commented:
> I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en
> 
> Changelog category (leave one):
> 
> * New Feature
> 
> Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
> 
> Add projection support for MergeTree* tables.
> 
> Detailed description / Documentation draft:
> 
> The documentation is almost listed in #14730
> 
> Possible TODOs.
> 
> * [ ]  Complete query routing for projections. Currently alias in group-by will not work.
> * [ ]  Normal projections with different order by clause
> * [ ]  ProjectionMergeTree to store projection data only
> * [ ]  Secondary indices via projection

@amosbird : 
This is a very valuable feature;
I have some questions about projects.
1. What is the difference between projections and materialized views?
2. Does projections also use the idea of kylin cube? Pre-aggregate the query results in advance. The aggregated results are directly queried during query. Projections: Are the combinations of all dimensions calculated in advance?

itzikiusa commented:
Long expected feature!
Waiting for PR to be approved, and the following steps, specially different order by
Thank you very much for your work!

KochetovNicolai committed:
Try to merge projectons faster.

KochetovNicolai committed:
Cleanup code.

KochetovNicolai committed:
Fix prewhere.

amosbird committed:
add metadata_snapshot to getQueryProcessingStage

amosbird committed:
Check if pipeline is simple and add more comments

KochetovNicolai committed:
Try to enable normal projections.

KochetovNicolai committed:
Disable normal projection by the number of granules.

KochetovNicolai committed:
Disable read-in-order optimization in case of projections.

KochetovNicolai committed:
Added test with normal projections.

amosbird committed:
Reformat and fix some tests

KochetovNicolai committed:
Add ProjectionDescription::Type enum. Remove some checks.

KochetovNicolai committed:
Remove copy constructor from ProjectionDescription

amosbird committed:
Use ActionsDAG to rewrite projection queries

amosbird committed:
Fix missing columns

amosbird committed:
Fix aggregation keys order

amosbird committed:
support prewhere, row_filter, read_in_order and decent projection selection

TODO set index analysis in projection

amosbird committed:
Rebase and fix

amosbird committed:
fix ya make (projection only)

amosbird committed:
Fix more

amosbird committed:
Fix tests

KochetovNicolai committed:
More tests.

amosbird committed:
Fix tests

amosbird committed:
Fix more tests

amosbird committed:
More fixes

amosbird committed:
Even more fixes

amosbird committed:
Remove projection type ast

amosbird commented:
```
2021-05-05 23:51:05 00992_system_parts_race_condition_zookeeper_long:                       Received exception from server (version 21.6.1):
2021-05-05 23:51:05 Code: 999. DB::Exception: Received from localhost:9000. DB::Exception: Run time inconsistency, path: /clickhouse/task_queue/ddl/query-0000003330. 
2021-05-05 23:51:06 Received exception from server (version 21.6.1):
2021-05-05 23:51:06 Code: 81. DB::Exception: Received from localhost:9000. DB::Exception: There was an error on [localhost:19000]: Code: 81, e.displayText() = DB::Exception: Database test_06rrwt doesn't exist (version 21.6.1.6765). 
2021-05-05 23:51:06 [ FAIL ] 4.53 sec. - return code 81
2021-05-05 23:51:06 Code: 81. DB::Exception: Received from localhost:9000. DB::Exception: Database `default` doesn't exist. 
2021-05-05 23:51:06 
2021-05-05 23:51:06 , result:
2021-05-05 23:51:06 
2021-05-05 23:51:06 
```

amosbird commented:
> Integration tests (asan) — Timeout, fail: 0, passed: 1314, flaky: 1
What does this mean?

amosbird committed:
Fix issues in normal projection

amosbird committed:
Comment debug info out

amosbird committed:
Fix

KochetovNicolai committed:
fix fetch for projection.

KochetovNicolai committed:
Remove unused code.

KochetovNicolai committed:
Try fix test.

KochetovNicolai committed:
Fix part check for projections.

KochetovNicolai commented:
Currently, most of mentioned in https://github.com/ClickHouse/ClickHouse/issues/14730#issue-698305277 is implemented. Exceptions are: `WHERE` in projection definition and everything from `Projection Analysis and Settings`.

This feature is experimental. Enable `allow_experimental_projection_optimization` to use projections in `SELECT`.
Projection with `ORDER BY` uses it's own primary key when selected. Read-in-order optimization also supported. 
Projection with `GROUP BY` uses stored pre-aggregated data.

Alter update/delete can't be used together with projections (currently).

There is still not enough tests. Also, not all the code was thoughtfully reviewed. However, I am going to merge this pr, because the feature works in general (and to avoid major conflicts).

amosbird commented:
https://clickhouse-test-reports.s3.yandex.net/20202/817bc1377c131e83d13bbd90260f8f2d97856bfd/integration_tests_(release)/integration_run_test_cluster_copier_test_py_1.txt

cluster copier test throws code 85: `FORMAT_IS_NOT_SUITABLE_FOR_INPUT`

amosbird commented:
https://clickhouse-test-reports.s3.yandex.net/20202/817bc1377c131e83d13bbd90260f8f2d97856bfd/integration_tests_(thread)/integration_run_test_jbod_balancer_test_py_0.txt

insert timeout

abyss7 commented on tests/queries/skip_list.json:
```diff
@@ -392,6 +392,9 @@
         "01475_read_subcolumns_storages",
         "01674_clickhouse_client_query_param_cte",
         "01666_merge_tree_max_query_limit",
+        "01710_projections",
```
Why don't you implement parsing for ANTLR grammar?

amosbird replied:
Done

alexey-milovidov commented:
@KochetovNicolai There is no comment about `MergeTreeDataSelectCache`.

You've missed the case when PVS-Studio argued about uncomprehensible code:
> PVS check — Have 7 new errors, 21 total errors 

https://clickhouse-test-reports.s3.yandex.net/20202/817bc1377c131e83d13bbd90260f8f2d97856bfd/pvs_studio_report/pvs-studio-html-report/sources/MergeTreeDataSelectExecutor.cpp_13.html#ln583

The code is correct but it becomes too tangled, which PVS-Studio does not like.

alexey-milovidov commented:
We need clear explanation - what MergeTreeDataSelectCache is doing, what's going on in the code...

olgarev commented:
Internal ticket DOCSUP-12894

zhouerlin commented:
Currently, projection does not support the where clause,
What is the difficulty in implementing it or for what purpose it was designed to do so?
